x/emissions/module/abci.go

```diff
@@ -49,6 +49,10 @@ func EndBlocker(ctx context.Context, am AppModule) error {
 		sdkCtx.Logger().Error("Error calculating global emission per topic: ", err)
 		return errors.Wrapf(err, "Rewards error")
 	}
+	err = am.keeper.ClearRewardableTopics(ctx)
```

Comment:
Sorry, I can't understand well why we clear rewardable topics at here. Please correct me if I am wrong.
We add rewardable topics at end of epoch. And this distribute the reward to these topic at next epoch. Because every end of epoch we get weights which is used for getting rewardable topic at the [code](https://github.com/allora-network/allora-chain/blob/db8fef09342a8d9b086ce75c5f3d89842c30fd30/x/emissions/module/rewards/rewards.go#L44). So these rewardable topics could be removed before the end of epoch, isn't it?
Please lmk if I misunderstood anything.
I will also test with this one

Reply:
Test looks good, thank you for that.

Not entirely sure what you're trying to say frankly, maybe you can rephrase? Will try to answer with what I understand:

Before this PR, we calc rewards per topic, then remove the rewardable topic, one at a time, all within EmitRewards. Then in rest of EndBlock we add topics to rewardable set.

This PR exists because in ^that setup, if there is ever an error in calculating rewards, as there was exactly once for each topic during the v0.4.0 migration, because participants scores were 0s, which was expected after clearing off all scores (=> there were div by 0 errors), rewardable topics was filled up, and topics are not being removed.

In this PR, we calculate rewards for all topics, then clear off all rewardable topics, then add rewardable topics (so rewards are calculated next block) as we find them (using active topic mgmt).

The sort you linked is still useful because it allows us to sort topics in a reliable way => nondeterminism via looping over maps.

Reply:
Update: @RedBird96 and I spoke and the side and he's totally right -- the weights are taken from the active set, and these weights are needed by each topic in rewardable set, but the rewardable set is considered after the topics are no longer active (the block after they're active, if they are reward-ready). This is being addressed in https://github.com/allora-network/allora-chain/pull/592